src/hotspot/share/runtime/objectMonitor.inline.hpp

```diff
@@ -135,63 +152,78 @@ inline void ObjectMonitor::set_recursions(size_t recursions) {
+         ", expected=" INT64_FORMAT, prev, old_value);
 #endif
   Atomic::store(&_owner, new_value);
   log_trace(monitorinflation, owner)("set_owner_from(): mid="
-                                     INTPTR_FORMAT ", old_value=" INTPTR_FORMAT
-                                     ", new_value=" INTPTR_FORMAT, p2i(this),
-                                     p2i(old_value), p2i(new_value));
+                                     INTPTR_FORMAT ", old_value=" INT64_FORMAT
+                                     ", new_value=" INT64_FORMAT, p2i(this),
+                                     old_value, new_value);
 }
 
-// Simply set _owner field to self; current value must match basic_lock_p.
-inline void ObjectMonitor::set_owner_from_BasicLock(void* basic_lock_p, JavaThread* current) {
-#ifdef ASSERT
-  void* prev = Atomic::load(&_owner);
-  assert(prev == basic_lock_p, "unexpected prev owner=" INTPTR_FORMAT
-         ", expected=" INTPTR_FORMAT, p2i(prev), p2i(basic_lock_p));
-#endif
-  // Non-null owner field to non-null owner field is safe without
-  // cmpxchg() as long as all readers can tolerate either flavor.
-  Atomic::store(&_owner, current);
-  log_trace(monitorinflation, owner)("set_owner_from_BasicLock(): mid="
-                                     INTPTR_FORMAT ", basic_lock_p="
-                                     INTPTR_FORMAT ", new_value=" INTPTR_FORMAT,
-                                     p2i(this), p2i(basic_lock_p), p2i(current));
+inline void ObjectMonitor::set_owner_from(int64_t old_value, JavaThread* current) {
+  set_owner_from_raw(old_value, owner_from(current));
 }
 
 // Try to set _owner field to new_value if the current value matches
 // old_value. Otherwise, does not change the _owner field. Returns
 // the prior value of the _owner field.
-inline void* ObjectMonitor::try_set_owner_from(void* old_value, void* new_value) {
-  void* prev = Atomic::cmpxchg(&_owner, old_value, new_value);
+inline int64_t ObjectMonitor::try_set_owner_from_raw(int64_t old_value, int64_t new_value) {
+  assert((int64_t)new_value < ThreadIdentifier::current(), "must be reasonable");
+  int64_t prev = Atomic::cmpxchg(&_owner, old_value, new_value);
   if (prev == old_value) {
     log_trace(monitorinflation, owner)("try_set_owner_from(): mid="
-                                       INTPTR_FORMAT ", prev=" INTPTR_FORMAT
-                                       ", new=" INTPTR_FORMAT, p2i(this),
-                                       p2i(prev), p2i(new_value));
+                                       INTPTR_FORMAT ", prev=" INT64_FORMAT
+                                       ", new=" INT64_FORMAT, p2i(this),
+                                       prev, new_value);
   }
   return prev;
 }
 
+inline int64_t ObjectMonitor::try_set_owner_from(int64_t old_value, JavaThread* current) {
+  return try_set_owner_from_raw(old_value, owner_from(current));
+}
+
+inline bool ObjectMonitor::has_successor() {
```

Comment:
Why are _succ accesses atomic here when previously they were not?

Reply:
General convention is that racily accessed variables should be accessed via Atomic::load/store to make it clear(er) they are racy accesses. But I agree it seems odd when direct accesses to `_succ` in the main cpp file are not atomic.

Reply:
> Why are _succ accesses atomic here when previously they were not?
>
They should had always been atomic.

Reply:
> But I agree it seems odd when direct accesses to _succ in the main cpp file are not atomic.
>
There was only one remaining direct access in debugging function `print_debug_style_on` which I fixed now.